src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -331,70 +339,78 @@ struct
       ( Protocol_state.Blockchain_state.ledger_hash blockchain_state
       , Protocol_state.Blockchain_state.staged_ledger_hash blockchain_state )
     in
-    let transitioned_staged_ledger =
+    let%bind _ = try_exn (fun () -> List.hd []) in
+    let%bind transitioned_staged_ledger =
       match
-          Inputs.Staged_ledger.apply ~logger:t.logger staged_ledger
-            (Inputs.External_transition.staged_ledger_diff transition)
-          |> Or_error.ok_exn
-      
+        Inputs.Staged_ledger.apply ~logger:t.logger staged_ledger
+          (Inputs.External_transition.staged_ledger_diff transition)
       with
-      | ( `Hash_after_applying staged_ledger_hash
-        , `Ledger_proof proof_opt
-        , `Staged_ledger transitioned_staged_ledger )
-      ->
-        let target_ledger_hash =
-          match proof_opt with
-          | None ->
-              Option.value_map
-                (Inputs.Staged_ledger.current_ledger_proof
-                   transitioned_staged_ledger)
-                ~f:ledger_hash_from_proof
-                ~default:
-                  (Frozen_ledger_hash.of_ledger_hash
-                     (Ledger.Db.merkle_root t.root_snarked_ledger))
-          | Some proof -> ledger_hash_from_proof proof
-        in
-        if
-          Frozen_ledger_hash.equal target_ledger_hash
-            blockchain_state_ledger_hash
-          && Staged_ledger_hash.equal staged_ledger_hash
-               blockchain_staged_ledger_hash
-        then transitioned_staged_ledger
-        else
-          failwith
-            "Snarked ledger hash and Staged ledger hash after applying the \
-             diff does not match blockchain state's ledger hash and staged \
-             ledger hash resp.\n"
+      | Ok
+          ( `Hash_after_applying staged_ledger_hash
+          , `Ledger_proof proof_opt
+          , `Staged_ledger transitioned_staged_ledger ) ->
+          let target_ledger_hash =
+            match proof_opt with
+            | None ->
+                Option.value_map
+                  (Inputs.Staged_ledger.current_ledger_proof
+                     transitioned_staged_ledger)
+                  ~f:ledger_hash_from_proof
+                  ~default:
+                    (Frozen_ledger_hash.of_ledger_hash
+                       (Ledger.Db.merkle_root t.root_snarked_ledger))
+            | Some proof -> ledger_hash_from_proof proof
+          in
+          if
+            Frozen_ledger_hash.equal target_ledger_hash
+              blockchain_state_ledger_hash
+            && Staged_ledger_hash.equal staged_ledger_hash
+                 blockchain_staged_ledger_hash
+          then Ok transitioned_staged_ledger
+          else
+            (*TODO: Punish*)
+            Error
+              (`Validation_error
```

Comment:
Can we collect all error cases that can be thrown by this in a single ADT? Then the handler can better handle some of the exception cases (such as `Parent_not_found`).